docker-storage-setup.sh

```diff
@@ -111,12 +115,48 @@ get_deferred_deletion_string() {
   [ -z "$major" ] && return 0
   [ -z "$minor" ] && return 0
 
-  # docker 1.9 onwards supports deferred device deletion. Enable it.
-  if [ $major -gt 1 ] || ([ $major -eq 1 ] && [ $minor -ge 9 ]);then
-    echo "--storage-opt dm.use_deferred_deletion=true"
+  if should_enable_deferred_deletion $major $minor; then
+     echo "--storage-opt dm.use_deferred_deletion=true"
   fi
 }
 
+should_enable_deferred_deletion() {
+   # docker 1.9 onwards supports deferred device deletion. Enable it.
+   local major=$1
+   local minor=$2
+   if [ $major -lt 1 ] || ([ $major -eq 1 ] && [ $minor -lt 9 ]);then
+      return 1
+   fi
+   if platform_supports_deferred_deletion; then
+      return 0
+   fi
+   return 1
+}
+
+platform_supports_deferred_deletion() {
+        local deferred_deletion_supported=1
+        trap cleanup_pipes EXIT
+        mkfifo $PIPE1
+        mkfifo $PIPE2
+        mount -o bind $TEMPDIR $TEMPDIR
+        if [ ! -x "/usr/bin/dss-child-read-write.sh" ];then
+           return 1
+        fi
+        unshare -m /usr/bin/dss-child-read-write.sh $PIPE1 $PIPE2 &
+        read -t 10 n <>$PIPE1
+        umount $TEMPDIR
+        rmdir $TEMPDIR > /dev/null 2>&1
+        deferred_deletion_supported=$?
+        echo "finish" > $PIPE2
```

Comment:
Given we introduced timeout logic, we will have to differentiate between whether we timed out or child sent us an message. If we timed out we want to return "1". Current code will return 0.

May be read the value of the variable after read and if it is "start", then we know child sent us a valid message.